### ISSUE TITLE
javascript: Got "ctags: Warning: ignoring null tag" 

### ISSUE THREAD

**Issue by masatake** [universal-ctags/ctags#900]
```
[yamato@x201]~/var/ctags-github% cat /tmp/foo.js 
if (true) {
    this[0] = function() {};
}

[yamato@x201]~/var/ctags-github% ./ctags -o - /tmp/foo.js 
ctags: Warning: ignoring null tag in /tmp/foo.js(line: 2)
[yamato@x201]~/var/ctags-github% js /tmp/foo.js 
[yamato@x201]~/var/ctags-github%
```

**Comment by cristianoliveira** [comment]
Is there any update on this?

**Comment by cristianoliveira** [comment]
Apparently it is related to Object destructuring like:

```js
    const { name, email } = user;

```

**Comment by masatake** [comment]
I have not inspected the warning well.
However, the input you gave looks valid as I tested the input on node command
I would like you to open a new issue for the input.
If you are good at JavaScript, I would like to know the specification as reference.
See  https://github.com/universal-ctags/ctags/issues/1949#issuecomment-457829914 how I struggled.

```
const { name, email } = user;
```
`name` and `email` should be tagged well.

```

[yamato@slave]~% node
> var user = {name: 'a', email: 'b'}
var user = {name: 'a', email: 'b'}
undefined
> const { name, email } = user;
const { name, email } = user;
undefined
> name
name
'a'
```

**Comment by cristianoliveira** [comment]
> If you are good at JavaScript, I would like to know the specification as reference.

Not sure if I got your question but the specification for the Object Destructuring is on the ECMAScript specification:
https://www.ecma-international.org/ecma-262/6.0/#sec-destructuring-assignment

:)

**Comment by masatake** [comment]
Thanks. That is what I need.

**Comment by winkee01** [comment]
Why does this issue not be addressed? I've got the same problem.

**Comment by winkee01** [comment]
> Thanks. That is what I need.

how did you remove the warning?

**Comment by masatake** [comment]
> how did you remove the warning?

I didn't remove the warning.

**Comment by winkee01** [comment]
> > how did you remove the warning?
> 
> I didn't remove the warning.

well, I've got the same issue, `{A, B} = some_obj` can be parsed correctly by ctags.

**Comment by winkee01** [comment]
Looks like no one is going to fix this issue

**Comment by masatake** [comment]
> Looks like no one is going to fix this issue

For the destructural binding, I tried once.

```
[jet@living]/tmp% cd /tmp
cd /tmp
[jet@living]/tmp% git clone https://github.com/masatake/ctags.git
git clone https://github.com/masatake/ctags.git
Cloning into 'ctags'...
...
[jet@living]/tmp% cd ctags
cd ctags
[jet@living]/tmp/ctags% git checkout  origin/js-destructual-binding 
git checkout  origin/js-destructual-binding 
Note: switching to 'origin/js-destructual-binding'.
...
HEAD is now at 4eeac5b9 JavaScript: support more forms of destructuring assignments
[jet@living]/tmp/ctags% bash ./autogen.sh
bash ./autogen.sh
+ type autoreconf
...
optlib2c: translating optlib/inko.ctags to optlib/inko.c
+ ./misc/optlib2c optlib/inko.ctags
+ exit 0
[jet@living]/tmp/ctags% ./configure
./configure
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
...
config.status: creating config.h
config.status: executing depfiles commands
[jet@living]/tmp/ctags% make -j 9
make -j 9
REPOINFO   main/repoinfo.h
make  all-recursive
make[1]: Entering directory '/tmp/ctags'
make[2]: Entering directory '/tmp/ctags'
...
  AR       libctags.a
  CCLD     ctags
  CCLD     mini-geany
make[2]: Leaving directory '/tmp/ctags'
make[1]: Leaving directory '/tmp/ctags'
[jet@living]/tmp/ctags% cat /tmp/bar.js 
cat /tmp/bar.js 
const { name, email } = user;

[jet@living]/tmp/ctags% ./ctags -o - /tmp/bar.js 
./ctags -o - /tmp/bar.js 
email	/tmp/bar.js	/^const { name, email } = user;$/;"	C
name	/tmp/bar.js	/^const { name, email } = user;$/;"	C
```

The branch introduces so many changes. 
So merging it to the master branch we use daily needs more consideration.

**Comment by masatake** [comment]
```
this[0] = function() {};
```

In this case, ctags says nothing.

```
{
	this[0] = function() {};
}
```

In this case, I got a warning. This will be an important hint fix for the original input.

**Comment by winkee01** [comment]
> > Looks like no one is going to fix this issue
> 
> For the destructural binding, I tried once.
> 
> ```
> [jet@living]/tmp% cd /tmp
> cd /tmp
> [jet@living]/tmp% git clone https://github.com/masatake/ctags.git
> git clone https://github.com/masatake/ctags.git
> Cloning into 'ctags'...
> ...
> [jet@living]/tmp% cd ctags
> cd ctags
> [jet@living]/tmp/ctags% git checkout  origin/js-destructual-binding 
> git checkout  origin/js-destructual-binding 
> Note: switching to 'origin/js-destructual-binding'.
> ...
> HEAD is now at 4eeac5b9 JavaScript: support more forms of destructuring assignments
> [jet@living]/tmp/ctags% bash ./autogen.sh
> bash ./autogen.sh
> + type autoreconf
> ...
> optlib2c: translating optlib/inko.ctags to optlib/inko.c
> + ./misc/optlib2c optlib/inko.ctags
> + exit 0
> [jet@living]/tmp/ctags% ./configure
> ./configure
> checking for a BSD-compatible install... /usr/bin/install -c
> checking whether build environment is sane... yes
> ...
> config.status: creating config.h
> config.status: executing depfiles commands
> [jet@living]/tmp/ctags% make -j 9
> make -j 9
> REPOINFO   main/repoinfo.h
> make  all-recursive
> make[1]: Entering directory '/tmp/ctags'
> make[2]: Entering directory '/tmp/ctags'
> ...
>   AR       libctags.a
>   CCLD     ctags
>   CCLD     mini-geany
> make[2]: Leaving directory '/tmp/ctags'
> make[1]: Leaving directory '/tmp/ctags'
> [jet@living]/tmp/ctags% cat /tmp/bar.js 
> cat /tmp/bar.js 
> const { name, email } = user;
> 
> [jet@living]/tmp/ctags% ./ctags -o - /tmp/bar.js 
> ./ctags -o - /tmp/bar.js 
> email	/tmp/bar.js	/^const { name, email } = user;$/;"	C
> name	/tmp/bar.js	/^const { name, email } = user;$/;"	C
> ```
> 
> The branch introduces so many changes.
> So merging it to the master branch we use daily needs more consideration.

I've tried it, a very good solution, thanks.
Did you made these branches, so many patches.

**Comment by dietrichm** [comment]
Hi

Although this is not contributing to the issue in ctags itself, I created a `ctags` shell function that ignores this specific warning for JavaScript files: https://gist.github.com/dietrichm/9e518549fd609173182ce36ac4726df2

Just in case you frequently run ctags on the command line and want to hide the warnings until the issue is fixed.

Dietrich

**Comment by Strangehill** [comment]
I also get this warning with array destructuring :
`const [lang, setLang] = useState("en")`

**Comment by masatake** [comment]
@Strangehill, could you add the same comment to  #1112?

**Comment by Strangehill** [comment]
of course

**Comment by masatake** [comment]
Thank you.

**Comment by WouterSioen** [comment]
FYI: If you pass the `--quiet=yes` command-line option, this warning is not shown anymore. While not being a perfect solution, it at least silences the clutter.

**Comment by coderj001** [comment]
I also got the same error 
[![https://imgur.com/OriFjrG.png](https://imgur.com/OriFjrG.png)](https://imgur.com/OriFjrG.png)

server.js
[![https://imgur.com/S1MC5X5.png](https://imgur.com/S1MC5X5.png)](https://imgur.com/S1MC5X5.png)

**Comment by AdrienGiboire** [comment]
> FYI: If you pass the `--quiet=yes` command-line option, this warning is not shown anymore. While not being a perfect solution, it at least silences the clutter.

That does not shut it up on my side, at all :shrug: 

I'll have to go with @dietrichm alternative for the time being as I'm heavily using deconstruction which ctags really spammy.

**Comment by AdrienGiboire** [comment]
> Could you try #3435?

It seems to be working to a limit.

Deconstruction in a multilines block of code fails:

```js
files.push({
  relative,
  basename: name,
  absolute: loc,
  mtime: +stat.mtime
});
```

I don't know if it is related but the following for fails as well:
```js
computed: {
  ...mapGetters([
    'getAsylumAssistant',
    'getModel',
    'getType',
  ]),
},
```

I thought these forms would work but apparently not:

```js
this.stagedlocalRecord = { localRecord, localRecordIndex }
// ---
return this.saveData({ partyId, id, data, rubyType: dbTypes[type], jsType: type })
// ---
createDocketEntryModel ({ commit, state }, requestData) {
  // ...
}
```

And there might be more but it takes too long for me to check all the warnings :sweat_smile:

**Comment by jafl** [comment]
For the original issue, the fix is in #3762.  Closing, since #3435 is working on the much larger destructuring issue.
